core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstringBefore.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-FileCopyrightText: none
+ * SPDX-License-Identifier: CC0-1.0
+ */
+
+package gov.nist.secauto.metaschema.core.metapath.function.library;
+
+import gov.nist.secauto.metaschema.core.metapath.DynamicContext;
+import gov.nist.secauto.metaschema.core.metapath.ISequence;
+import gov.nist.secauto.metaschema.core.metapath.MetapathConstants;
+import gov.nist.secauto.metaschema.core.metapath.function.FunctionUtils;
+import gov.nist.secauto.metaschema.core.metapath.function.IArgument;
+import gov.nist.secauto.metaschema.core.metapath.function.IFunction;
+import gov.nist.secauto.metaschema.core.metapath.item.IItem;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IDecimalItem;
+import gov.nist.secauto.metaschema.core.metapath.item.atomic.IStringItem;
+import gov.nist.secauto.metaschema.core.util.ObjectUtils;
+
+import java.util.List;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Implements <a href=
+ * "https://www.w3.org/TR/xpath-functions-31/#func-substring-before">fn:substring-before</a>.
+ */
+public final class FnSubstringBefore {
+  private static final String NAME = "substring-before";
+  @NonNull
+  static final IFunction SIGNATURE_TWO_ARG = IFunction.builder()
+      .name(NAME)
+      .namespace(MetapathConstants.NS_METAPATH_FUNCTIONS)
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("arg1")
+          .type(IStringItem.class)
+          .zeroOrOne()
+          .build())
+      .argument(IArgument.builder()
+          .name("arg2")
+          .type(IStringItem.class)
+          .zeroOrOne()
+          .build())
+      .returnType(IStringItem.class)
+      .returnOne()
+      .functionHandler(FnSubstringBefore::executeTwoArg)
+      .build();
+
+  private FnSubstringBefore() {
+    // disable construction
+  }
+
+  @SuppressWarnings({ "unused", "PMD.OnlyOneReturn" })
+  @NonNull
+  private static ISequence<IStringItem> executeTwoArg(
+      @NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+
+	// From the XPath 3.1 specification:
+	// If the value of $arg1 or $arg2 is the empty sequence, or contains only
+	// ignorable collation units, it is interpreted as the zero-length string.
+    IStringItem arg1 = arguments.get(0) == ISequence.empty() ? IStringItem.valueOf("") : FunctionUtils.asTypeOrNull(arguments.get(0).getFirstItem(true));
```

Comment:
Using `==` to compare two objects is [not ideal in Java](https://pmd.github.io/pmd/pmd_rules_java_errorprone.html#compareobjectswithequals).
```suggestion
    IStringItem arg1 = arguments.get(0).empty() ? IStringItem.valueOf("") : FunctionUtils.asTypeOrNull(arguments.get(0).getFirstItem(true));
```

Reply:
I think we are looking for `isEmpty()` based on IDE warnings, but I am pushing it up in a minute on this PR and the `-after` one as well to wrap up reviews.

Reply:
I pushed it up with `.isEmpty()` in https://github.com/metaschema-framework/metaschema-java/pull/194/commits/3df44891d412775d898bc9852fa35910b0ed1425, resolving this review item.